sidecarhttp/http_api.go

```diff
@@ -0,0 +1,310 @@
+		sendJsonError(response, 404, "Not Found - Invalid content type extension")
+		return
+	}
+
+	response.Header().Set("Content-Type", "application/json")
+
+	var listMembers []*memberlist.Node
+	var clusterName string
+	if s.list != nil {
+		listMembers = s.list.Members()
+		sort.Sort(listByName(listMembers))
+		clusterName = s.list.ClusterName()
+	}
+
+	members := make(map[string]*ApiServer, len(listMembers))
+
+	var jsonBytes []byte
+	var err error
+
+	func() { // Wrap critical section
+		s.state.RLock()
+		defer s.state.RUnlock()
+
+		for _, member := range listMembers {
+			if s.state.HasServer(member.Name) {
+				members[member.Name] = &ApiServer{
+					Name:         member.Name,
+					LastUpdated:  s.state.Servers[member.Name].LastUpdated,
+					ServiceCount: len(s.state.Servers[member.Name].Services),
+				}
+			} else {
+				members[member.Name] = &ApiServer{
+					Name:         member.Name,
+					LastUpdated:  time.Unix(0, 0),
+					ServiceCount: 0,
+				}
+			}
+		}
+
+		result := ApiServices{
+			Services:       s.state.ByService(),
+			ClusterMembers: members,
+			ClusterName:    clusterName,
+		}
+
+		jsonBytes, err = json.MarshalIndent(&result, "", "  ")
+	}()
+
+	if err != nil {
+		log.Errorf("Error marshaling state in servicesHandler: %s", err.Error())
+		sendJsonError(response, 500, "Internal server error")
+		return
+	}
+
+	response.Write(jsonBytes)
+}
+
+// stateHandler simply dumps the JSON output of the whole state object. This is
+// useful for listeners or other clients that need a full state dump on startup.
+func (s *SidecarApi) stateHandler(response http.ResponseWriter, req *http.Request, params map[string]string) {
+	defer req.Body.Close()
+
+	s.state.RLock()
+	defer s.state.RUnlock()
+
+	if params["extension"] != "json" {
+		sendJsonError(response, 404, "Not Found - Invalid content type extension")
+		return
+	}
+
+	response.Header().Set("Content-Type", "application/json")
+	response.Header().Set("Access-Control-Allow-Origin", "*")
+	response.Header().Set("Access-Control-Allow-Methods", "GET")
+	response.Write(s.state.Encode())
+	return
+}
+
+// Send back a JSON encoded error and message
+func sendJsonError(response http.ResponseWriter, status int, message string) {
+	output := map[string]string{
+		"status":  "error",
+		"message": message,
+	}
+
+	jsonBytes, err := json.Marshal(output)
+
+	if err != nil {
+		log.Errorf("Error encoding json error response: %s", err.Error())
+		response.WriteHeader(500)
+		response.Write([]byte("Interval server error"))
+		return
+	}
+
+	response.Header().Set("Content-Type", "application/json")
+	response.WriteHeader(status)
+	response.Write(jsonBytes)
+}
+
+func wrap(fn func(http.ResponseWriter, *http.Request, map[string]string)) http.HandlerFunc {
+	return func(response http.ResponseWriter, req *http.Request) {
+		fn(response, req, mux.Vars(req))
+	}
+}
+
+// Used by the servicesHandler to sort Memberlist cluster nodes
+type listByName []*memberlist.Node
```

Comment:
Why duplicate [this code](https://github.com/Nitro/sidecar/pull/15/files#diff-ce823b4b938ef73330d3b1efef576030R89)?

Reply:
I moved it there but forgot to remove it here :) Fixing.